adapters/from_sql.py

```diff
@@ -0,0 +1,194 @@
+
+  Args:
+    release_data: the raw release data from SQL (tuple)
+    airflow_db: the airflow database connection object
+
+  Returns:
+    release_objects
+  """
+  release_named_tuple = namedtuple('Row', ['dag_id', 'execution_date'])
+  release_objects = {}  # resources.py expects a dict of objects
+  for item in release_data:  # iterate through each release in release_data
+    item = release_named_tuple._make(item)
+    release = Release()  # initialize the release object
+    started = to_timestamp(item.execution_date)
+    task_ids, most_recent_task, state = get_task_info(item.dag_id, started, airflow_db)
+    xcom_dict, green_sha = get_xcom(started, item.dag_id, airflow_db)
+    release.release_id = item.dag_id + '@' + str(item.execution_date)
+    release.tasks = task_ids
+    release.started = started
+    release.links = construct_links(green_sha)  # TODO(dommarques) these need to be implemented into airflow first, or we make our own way to get the links pylint: disable=line-too-long
+    release.labels = [item.dag_id]
+    release.state = state
+    release.branch, release.release_type = parse_dag_id(item.dag_id)
+    if xcom_dict is None:
+      # allows for continuation, even if xcom has not been generated yet
+      release.name = release.release_id
+    else:
+      release.name = xcom_dict['VERSION']
+    if most_recent_task:
+      # allows for continuation even if tasks have not begun/been generated yet
+      release.last_modified = to_timestamp(most_recent_task.last_modified)
+      release.last_active_task = most_recent_task.task_name
+    else:
+      release.last_modified = to_timestamp(item.execution_date)
+      release.last_active_task = ''
+    release_objects[release.release_id] = release
+
+  return release_objects
+
+
+def read_tasks(task_data):
+  """Reads the tasks, transforms them into objects.
+
+  Args:
+    task_data: a tuple of tuples with task info
+
+  Returns:
+    task_objects
+  """
+  task_named_tuple = namedtuple('Row', ['task_id', 'dag_id', 'execution_date', 'start_date', 'end_date', 'state'])
+  task_objects = []
+  for item in task_data:
+    task = Task()
+    logging.debug(item)
+    item = task_named_tuple._make(item)
+    task.task_name = item.task_id
+    task.add_dependency = None  # TODO(dommarques): figure out how to get this - dag info?
+    if item.start_date == None:
+      task.started = to_timestamp(item.execution_date)
+    else:
+      task.started = to_timestamp(item.start_date)
+    if item.state == None:
+      task.status = STATE_FROM_STRING.get('none')
+    else:
+      task.status = STATE_FROM_STRING.get(item.state)
+    if item.end_date == None:
+      task.last_modified = to_timestamp(item.execution_date)
+    else:
+      task.last_modified = to_timestamp(item.end_date)
+    task.log_url = 'https://youtu.be/dQw4w9WgXcQ'  # TODO(dommarques): figure out how to get the log in here
+    if item.state == None:
+      item.state == 'none'
+    else:
+      task.error = item.state
```

Comment:
do you want task.error to be set if state is successful  or pending ?

Reply:
It's totally up to you. I originally had it there just to make sure that the task error was always reading, but since we know it works now it doesn't need to be that way anymore

Reply:
I think task error should be a summary of the error. Do we have this data in Airflow ? @rkpagadala